Add adapter e3oncan to latest

MyHomeMyData opened the pull request.

MyHomeMyData committed:
Added adapter e3oncan

MyHomeMyData commented:
RE-CHECK!

mcm1957 commented:
Thanks for spending your time and providing a new adapter for ioBroker.

Your adapter will get a manual review as soon as possible. Please stand by - this might last one or two weeks. Feel free to continue your work and create new releases. You do NOT need to close or update this PR in case of new releases.

You will find the results of the review and evntually issues / suggestings as a communt to this PR. So please keep this PR watched. 

If you have any urgent questions feel free to ask.

mcm1957

mcm1957 commented:
reminder 24.01.2024

mcm1957 commented:
RE-CHECK!

MyHomeMyData commented:
Hallo Martin,

 

ich möchte Dir ein paar Hintergrundinfo zum Adapter e3oncan <https://github.com/MyHomeMyData/ioBroker.e3oncan.git>  geben.

 

Es geht hier darum, direkten Zugriff auf die Daten der neuesten Generation (E3) von Viessmann-Geräten (Wärmepumpen, Gasthermen, PV-Umrichter und -Speicher, usw.) zu bekommen. Dazu wird der CAN-Bus der Geräte genutzt.

 

Viessmann bietet auch eine API an, dafür gibt es bereits einen Adapter. Die API ist aber in der kostenlosen Version im Bereich Heizung mittelprächtig, im Bereich PV völlig nutzlos. Die Bezahlvariante kostet zwischen 8 und 20 (!) Euro im Monat und liefert trotzdem nur ein lückenhaftes Bild, ist wenig stabil und grausam zu bedienen.

 

Das hat mich und etliche andere Nutzer dieser Geräte massiv geärgert – zumal man oft mehrere zehntausend Euro für die Anlage ausgegeben hat. Deshalb haben sich Anfang 2023 ein paar Leute zusammengefunden und per Reverse Engineering und mit Hilfe öffentlich zugänglicher Quellen das Projekt open3e <https://github.com/open3e/open3e.git>  auf die Beine gestellt. Das ist eine Python-Anwendung die die Daten vom CAN-Bus abgreift, per MQTT bereitstellt und ohne Internetverbindung auskommt. Es gibt einen umfangreichen Thread im Viessmann-Forum dazu. Dieser Beitrag <https://community.viessmann.de/t5/Konnektivitaet/CAN-Bus-Home-Automation-E3-Generation-lokal-und-kostenlos/td-p/356066>  hat mit großem Abstand die meistens Likes im gesamten Viessmann-Forum! Viessmann unterstützt uns leider nicht, das haben wir auf verschiedenen Ebenen probiert. Sie lassen uns bisher aber zumindest in Ruhe. Nachdem sich Haier die Finger verbrannt hat, bleibt das hoffentlich auch so.

 

Ich war von Anfang an beim Projekt dabei und habe auf dieser Basis nun den Adapter entwickelt. Der Adapter läuft unabhängig von open3e. Es wird über einen USB2CAN-Adapter eine direkte Verbindung zum CAN-Bus der Anlage hergestellt. Darüber läuft die gesamte Kommunikation.

 

Die Viessmann Geräte können untereinander vernetzt sein. Sie tauschen dann viele Infos über den CAN-Bus aus. Deshalb hat der Adapter einen Modus, der die Kommunikation belauscht und relevante Daten abgreift. Ich nenne das „Collect“.

 

Zusätzlich gibt es die Möglichkeit über das Protokoll UDSonCAN <https://www.csselectronics.com/pages/uds-protocol-tutorial-unified-diagnostic-services>  (Universal Diagnostic Services on CAN) gezielt Daten abzufragen und zu verändern. Ich nenne das „Uds“.

 

Insgesamt gibt es mehr als 1000 Datenpunkte in den Geräten, mehrere hundert davon kann man verändern, z.B. die Zeitpläne für die Heizungs-Modi oder die Solltemperatur für das Warmwasser.

 

Beim ersten Start des Adapter sollten zwei Scan-Vorgänge durchgeführt werden: 1. Scan der vorhandenen Geräte auf dem Bus. 2. Scan sämtlicher Datenpunkte auf den einzelnen Geräten.

 

Um den Adapter ausprobieren zu können, braucht man eigentlich eine Viessmann-Anlage. Im Rahmen des open3e-Projekts ist aber auch ein Simulator für die E3-Geräte entstanden. Man kann einen virtuellen CAN (vcan0) einrichten und den Adapter mit dem Simulator koppeln. Damit kann man alles ausprobieren. Mit dieser Konstellation habe ich auch sehr viel getestet. Eine für diesen Zweck optimierte Version stelle ich in einem Fork <https://github.com/MyHomeMyData/vitualE3.git>  bereit. Da gibt’s auch eine Installationsanleitung.

 

So, ich hoffe, das war jetzt nicht zu viel Info. Falls Du Fragen hat, gerne melden.

 

Ich habe noch ein Thema, da bin ich unsicher: Für die Kommunikation zu den verschiedenen Geräten instanziiere ich jeweils eine Klasse. Beim Stopp des Adapters müssen die alle gestoppt werden und das kann einen Moment dauern. Deshalb habe ich die onUnload() Routine als async deklariert und warte dann, bis die Stopp-Routinen durch sind. Im Template war da kein async, deshalb bin ich unsicher, ob das so ok ist. Es scheint zumindest problemlos zu funktionieren.

 

Viele Grüße

Jürgen 

 

Von: Martin M. ***@***.***> 
Gesendet: Mittwoch, 17. Januar 2024 19:59
An: ioBroker/ioBroker.repositories ***@***.***>
Cc: MyHomeMyData ***@***.***>; Author ***@***.***>
Betreff: Re: [ioBroker/ioBroker.repositories] Add adapter e3oncan to latest (PR #3215)

 

Thanks for spending your time and providing a new adapter for ioBroker.

Your adapter will get a manual review as soon as possible. Please stand by - this might last one or two weeks. Feel free to continue your work and create new releases. You do NOT need to close or update this PR in case of new releases.

You will find the results of the review and evntually issues / suggestings as a communt to this PR. So please keep this PR watched.

If you have any urgent questions feel free to ask.

mcm1957

—
Reply to this email directly, view it on GitHub <https://github.com/ioBroker/ioBroker.repositories/pull/3215#issuecomment-1896450930> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/BCR4KA73I6S2QA735XQRJDDYPANNPAVCNFSM6AAAAABB6Q5652VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOJWGQ2TAOJTGA> .
You are receiving this because you authored the thread.  <https://github.com/notifications/beacon/BCR4KA4Q2CMHVSDK6YUE4PLYPANNPA5CNFSM6AAAAABB6Q5652WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTRBGFXE.gif> Message ID: ***@***.*** ***@***.***> >

mcm1957 commented:
First of all - THANK YOU for the time and effort you spend to maintain this adapter.

I would like to give some feedback based on my personal oppinion. @Apollon77 might have additional suggestions or even a different oppinion to one or the other statement. Please feel free to contact him if you cannot follow my suggestions or want to discuss some special aspects.

- [ ] invalid characters should be filtered from object ids
  
  Some characters are not allowed to be used as part of an object id. If an object id is not hardcoded or even depending on user input you should ensure that such an object id does not contain an invalid character.  Iobroker provides the constant  ```adapter.FORBIDDEN_CHARS```, i.e. by using the following snippet:

  ```
  function name2id(pName) {
    return (pName || '').replace(adapter.FORBIDDEN_CHARS, '_');
  }
  ```

  Please ensure that ALL state ids are filtered, either nearby creation / setting or during retrieval of data. Even data read from configuration should be filtered to avoid illegal characters passing in this way. (Filtering in jsonConfig is not sufficient as users could write the config object directly)

- [ ] reevaluate state roles

  Only the values specified here (https://github.com/ioBroker/ioBroker.docs/blob/master/docs/en/dev/stateroles.md) may be used as state roles. Do not invent new names as this might disturb the functionalyity of other adapters or vis.

  In addition the roles MUST match the read/write functionality. As an example a role value.* requires that the state is a readable state. Input states (write only) should have roles like level.*. Please read the description carefully. States with role 'button' should (must) have attribute 'common.read=false' set. A button ( "Taster" in german only triggers when you press but else has no state), so it should also be not readable. This is also like HomeMatic does it. A switch has clear states true or false and so can be read.

  Please avoid using general roles (state, value) whnever a dedicated role exists **although it might not be possiblewith this adapter**. 

  So consider this issue only as a reminder. If all states are correctly marked as write can be checked by yiou only.
  
- [ ] consider using adapter.setTimeout / this.setTimeout instead of (standard) setTimeout
  
  The adapter package provides wrapper routines for native setTimeout, setInterval, clearTimeout and clearInterval. Using those routines ensures that timers are cancelled on on load. Additional checks on those limits might be performed, too. So consider replacing native setTimeout/clearTimeout by adapter.setTimeout/adapter.clearTimeout or this.setTimeout/this.clearTimeout. The same refers to setInterval/clearInterval.

  In any case all started timers / intervalls must be cancelled at onUnload routine.

  In addtion consider replacing private sleep routine by this.delay provided by adapter core. This would solve the problem, that (at least currently) you sleep timers are not cancelled

- [ ] async onUnload   
  
  Ii've checked with core team. js-controller waits maximum of 500ms or until callback is called. Declaring onUnload routine as async is ok. Please note that your cleanup MUST be finished within 500ms. If this cannot be accomplished, the timeout could be increased but this is discuraged. Please discuss with Apollon77 if you feel a need to do.

- [ ] consider enabled:false

  Readme tells that an instance should not be started until configuration has been performed. You might consider setting 'common.enabled = false' at io-package.json. The defaultis to start a new instance immidiatly after it s created.

Thanks for reading and evaluating this suggestions.
McM1957

**Please add a comment when you have reviewed  the suggestions or at least commented the suggestions and you think the adapter is ready**

mcm1957 commented:
remember 8.2.2024

MyHomeMyData commented:
@mcm1957 Thank you very much for doing a review and for your valuable feedback. I went through the findings and here are my comments. All fixes/improvements are included to V0.6.16.

* _invalid characters should be filtered from object ids_
I added the suggested replace() statement to all relevant state operations

* _reevaluate state roles_
I changed/removed all non-conform role definitions. However, for some states I could not find a role definition other than "state" meeting all conditions.

*  _consider using adapter.setTimeout / this.setTimeout instead of (standard) setTimeout_
Now only this.setTimeout is used througout the adapter.

* _async onUnload_
After some optimizations I did several measurements of time to complete onUnload(). I tried to use a worst case configuration for those measurments. On a raspberry pi 4 I got values below 100 ms in about 80% of my tests. The largest value I got was 205 ms. So all values are well below 500 ms.
From my point of view this is fine and I would not plan/require additonal measures.

* _consider enabled:false_
This is a misunderstanding  caused by a unclear hint given in Readme. I changed the wording in Readme to - hopefully - make it more clear. The adapter needs to run during configuration. The hint in Readme is, to make sure **no other client** for UDSonCAN communication is running in parallel during configuration.

Pls. re-review the adapter based on V0.6.16 and give me your feedback. Thank you.

Best regards
Jürgen

mcm1957 commented:
RE-CHECK!

mcm1957 commented:
This adapter has been released to latest repository and should be visible within 24h maximum.

Please create a thread at https://forum.iobroker.net/category/91/tester titled like "Test Adapter " to collect some user feedback and provide a link to this topic when requesting addition to stable repository later.

Note: If an other testing topic already exists, it' OK to continue using this topic too.
